File: Question3-Python Socket Programming /ServerSocket.py
import socket

def bar_to_atmosphere(pressure_bar):
    return pressure_bar * 0.986923

def handle_client(client_socket):
    data = client_socket.recv(1024).decode('utf-8')
    try:
        pressure_bar = float(data)
        atmosphere_standard = bar_to_atmosphere(pressure_bar)
        client_socket.send(str(atmosphere_standard).encode('utf-8'))
    except ValueError:
        client_socket.send("Invalid input. Please provide a valid pressure value in bar.".encode('utf-8'))
    client_socket.close()

def main():
    host = '0.0.0.0'  
    port = 4040  

    server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    server_socket.bind((host, port))
    server_socket.listen(5)

    print("Server is listening on {}:{}".format(host, port))

    while True:
        client_socket, addr = server_socket.accept()
        print("Connection from:", addr)
        handle_client(client_socket)

if __name__ == "__main__":
    main()

File: Question2-C Socket Programming /ServerSocket.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <arpa/inet.h>

#define PORT 8443

int main() {
    // Create a socket
    int server_socket = socket(AF_INET, SOCK_STREAM, 0);
    if (server_socket < 0) {
        perror("Socket creation error");
        exit(EXIT_FAILURE);
    }

    // Bind the socket to an IP address and port
    struct sockaddr_in server_addr;
    server_addr.sin_family = AF_INET;
    server_addr.sin_addr.s_addr = INADDR_ANY;
    server_addr.sin_port = htons(PORT);

    if (bind(server_socket, (struct sockaddr *)&server_addr, sizeof(server_addr)) < 0) {
        perror("Binding error");
        close(server_socket);
        exit(EXIT_FAILURE);
    }

    // Listen for incoming connections
    if (listen(server_socket, 5) < 0) {
        perror("Listening error");
        close(server_socket);
        exit(EXIT_FAILURE);
    }

    printf("Server is listening on port %d...\n", PORT);

    while (1) {
        // Accept a client connection
        struct sockaddr_in client_addr;
        int client_socket;
        socklen_t client_addr_len = sizeof(client_addr);
        client_socket = accept(server_socket, (struct sockaddr *)&client_addr, &client_addr_len);
        if (client_socket < 0) {
            perror("Accepting error");
            close(server_socket);
            exit(EXIT_FAILURE);
        }

        // Generate a random number ranging from 100 to 999
        int random_number = rand() % 900 + 100;
        printf("Generated random number: %d\n", random_number);

        // Send the random number to the client
        send(client_socket, &random_number, sizeof(random_number), 0);

        // Close the connection
        close(client_socket);
    }

    return 0;
}
